### PR TITLE
[REF] mrp_subcontracting

### PR DESCRIPTION
**Pull request by amoyaux**
This PR modify the subcontracting process:
- Support extra quantity
- No more interaction with the subcontract order. Everything is directly accessible on the operation
- Dropshipping with subcontracting
- Bypass reservation process
- The partner type subcontractor is removed and every partner could be set on the BoM
- Update the initial demand will trigger the rules.
- Demo data
- Subcontractor is required on the BoM (no more BoM for all partners of type subcontractor)

Visually:
- Remove the stat button for linked operation.
- Register Components is only available when a subcontract component is tracked
- Button on operations in order to correct tracked components
- Show Operation button will ask to record components until the total quantity is provided(correction on finish lines are allowed after)

More details available in each commit
Task 1962187

### CODE REVIEW

File: addons/mrp/models/stock_warehouse.py [hunk 1]
```diff
@@ -101,6 +103,8 @@ def _get_route_name(self, route_type):
 
     def _get_global_route_rules_values(self):
         rules = super(StockWarehouse, self)._get_global_route_rules_values()
```
Comment (bealdav) [inline addons/mrp/models/stock_warehouse.py]: Please as you know `s / super(StockWarehouse, self). / super().`

and in all code below. Thanks

File: addons/mrp_subcontracting/__init__.py [hunk 2]
```diff
@@ -1,25 +1,4 @@
 # -*- coding: utf-8 -*-
```
Comment (bealdav) [inline addons/mrp_subcontracting/__init__.py]: As you know utf8 is the default in Py3

File: addons/mrp_subcontracting/models/res_company.py [hunk 3]
```diff
@@ -9,24 +9,34 @@ class ResCompany(models.Model):
 
     subcontracting_location_id = fields.Many2one('stock.location')
 
+    @api.model
+    def create_missing_subcontracting_location(self):
+        company_without_subcontracting_loc = self.env['res.company'].search(
+            [('subcontracting_location_id', '=', False)])
+        for company in company_without_subcontracting_loc:
+            company._create_subcontracting_location()
```
Comment (kos94ok-3D) [inline addons/mrp_subcontracting/models/res_company.py]: Why loop recordset and call function per record?

File: addons/mrp/wizard/change_production_qty.py [hunk 4]
```diff
@@ -35,8 +35,8 @@ def _update_finished_moves(self, production, qty, old_qty):
         modification = {}
         for move in production.move_finished_ids.filtered(lambda m: m.state not in ('done', 'cancel')):
             qty = (qty - old_qty) * move.unit_factor
-            modification[move] = (move.product_uom_qty - qty, move.product_uom_qty)
-            move[0].write({'product_uom_qty': move.product_uom_qty - qty})
+            modification[move] = (move.product_uom_qty + qty, move.product_uom_qty)
+            move[0].write({'product_uom_qty': move.product_uom_qty + qty})
```
Comment (sle-odoo) [inline addons/mrp/wizard/change_production_qty.py]: it would be nice to add a test for this

File: addons/mrp_subcontracting/models/stock_move.py [hunk 5]
```diff
@@ -3,13 +3,36 @@
...
+        compute='_compute_subcontract_move_line_ids',
+        inverse='_inverse_subcontract_move_line_ids',
+        string='Subcontracted Components', readonly=False)
+
+    def action_show_details(self):
+        action = super(StockMove, self).action_show_details()
+        action['context'].update({
+            'show_lots_m2o': True,
+            'show_lots_text': False,
+        })
```
Comment (sle-odoo) [inline addons/mrp_subcontracting/models/stock_move.py]: you shouldn't do that, or guard be if self.is_subcntract

File: addons/mrp/wizard/mrp_product_produce.py [hunk 6]
```diff
@@ -16,8 +16,12 @@ class MrpProductProduce(models.TransientModel):
...
     def default_get(self, fields):
         res = super(MrpProductProduce, self).default_get(fields)
-        if self._context and self._context.get('active_id'):
-            production = self.env['mrp.production'].browse(self._context['active_id'])
+        production = self.env['mrp.production']
+        if self.env.context and self.env.context.get('default_production_id'):
+            production = self.env['mrp.production'].browse(self.env.context['default_production_id'])
+        if not production and self.env.context and self.env.context.get('active_id'):
+            production = self.env['mrp.production'].browse(self.env.context['active_id'])
+        if production.exists():
```
Comment (sle-odoo) [inline addons/mrp/wizard/mrp_product_produce.py]:
``` python
production = self.env['mrp.production']
production_id = self.env.context.get('default_production_id') or self.env.context.get('active_id')
if production_id:
    production = self.env['mrp.production'].browse(production_id)
if production.exists():
```

File: addons/mrp_subcontracting/models/mrp_bom.py [hunk 7]
```diff
@@ -8,13 +8,9 @@ class MrpBom(models.Model):
...
 
     def _bom_subcontract_find(self, product_tmpl=None, product=None, picking_type=None, company_id=False, bom_type='subcontract', subcontractor=False):
         domain = self._bom_find_domain(product_tmpl=product_tmpl, product=product, picking_type=picking_type, company_id=company_id, bom_type=bom_type)
-        if subcontractor:
-            domain += ['|', ('subcontractor_ids', 'in', subcontractor.id), ('subcontractor_ids', '=', False)]
-        else:
-            domain += [('subcontractor_ids', '=', False)]
-        return self.search(domain, order='subcontractor_ids, sequence, product_id', limit=1)
-
+        bom = self.search(domain, order='subcontractor_ids, sequence, product_id').filtered(lambda b: subcontractor in b.subcontractor_ids)
```
Comment (sle-odoo) [inline addons/mrp_subcontracting/models/mrp_bom.py]: ('subcontractor_ids', 'in', subcontractor.id) should work

File: addons/mrp_subcontracting/models/stock_location.py [hunk 8]
```diff
@@ -0,0 +1,14 @@
...
+
+
+class StockLocation(models.Model):
+    _inherit = 'stock.location'
+
+    def should_bypass_reservation(self):
+        should_bypass_reservation = super(StockLocation, self).should_bypass_reservation()
+        if not should_bypass_reservation and self == self.env.company.subcontracting_location_id:
+            return True
+        return should_bypass_reservation
```
Comment (sle-odoo) [inline addons/mrp_subcontracting/models/stock_location.py]: won't work in multi company (access of two companies and subcontract amother
``` python
res = super(StockLocation, self).should_bypass_reservation()
if not should_bypass_reservation:
    self in self.env.company_ids.mapped('subcontracting_location_id)
return res
```
+ what if a company is forced to a partner and is not part of the property?

Comment (sle-odoo) [reply addons/mrp_subcontracting/models/stock_location.py]: to remove and override _action_assign, extract
```
            if move.location_id.should_bypass_reservation()\
                    or move.product_id.type == 'consu':
```
 to an helper

File: addons/mrp_subcontracting/models/stock_move.py [hunk 9]
```diff
@@ -3,39 +3,141 @@
...
+        action = super(StockMove, self).action_show_details()
+        if self.is_subcontract:
+            action['views'] = [(self.env.ref('stock.view_stock_move_operations').id, 'form')]
+            action['context'].update({
+                'show_lots_m2o': self.has_tracking != 'none',
+                'show_lots_text': False,
+            })
+        return action
+
+    def action_show_subonctract_details(self):
```
Comment (sle-odoo) [inline addons/mrp_subcontracting/models/stock_move.py]: typo

File: addons/mrp_subcontracting/models/stock_move.py [hunk 10]
```diff
@@ -3,39 +3,141 @@
...
+            if not move._has_tracked_subcontract_components():
+                continue
+            move.show_details_visible = True
+        return res
+
+    def _compute_show_subcontracting_details_visible(self):
+        for move in self:
+            if move.is_subcontract and move._has_tracked_subcontract_components() and\
+                    not float_is_zero(move.quantity_done, precision_rounding=move.product_uom.rounding):
+                move.show_subcontracting_details_visible = True
```
Comment (sle-odoo) [inline addons/mrp_subcontracting/models/stock_move.py]: assign the falsy values too for new ~new orm

File: addons/mrp_subcontracting/models/stock_move.py [hunk 11]
```diff
@@ -3,39 +3,141 @@
...
+        return any(m.has_tracking != 'none' for m in self.move_orig_ids.production_id.move_raw_ids)
+
+    def _update_subcontract_order_qty(self):
+        for move in self:
+            production = move.move_orig_ids.production_id
+            if production:
+                self.env['change.production.qty'].with_context(skip_activity=True).create({
+                    'mo_id': production.id,
+                    'product_qty': move.product_uom_qty
+                }).change_prod_qty()
```
Comment (sle-odoo) [inline addons/mrp_subcontracting/models/stock_move.py]: for all these new methods

Private attributes (_name, _description, _inherit, …)
Default method and _default_get
Field declarations
Compute, inverse and search methods in the same order as field declaration
Selection method (methods used to return computed values for selection fields)
Constrains methods (@api.constrains) and onchange methods (@api.onchange)
CRUD methods (ORM overrides)
Action methods
And finally, other business methods.

File: addons/mrp_subcontracting/models/stock_picking.py [hunk 12]
```diff
@@ -18,96 +19,107 @@ def _compute_display_action_record_components(self):
...
     # Action methods
     # -------------------------------------------------------------------------
+
+    def action_cancel(self):
+        for picking in self:
+            picking._get_subcontracted_productions()._action_cancel()
+        return super(StockPicking, self).action_cancel()
+
     def action_done(self):
+        res = super(StockPicking, self.with_context(skip_mo=True)).action_done()
```
Comment (sle-odoo) [inline addons/mrp_subcontracting/models/stock_picking.py]: you need to explain the skip_mo context key (dont_create_subcontract_order)

maybe split self in two recordset: subcontract and not subcontract

``` python
is_subcontract = {picking: any(move.is_subcontract for move in picking.move_lines) for picking in self}
for picking in self:
    if not is_subcontract.get(picking):
        continue
   production = move.move_orig_ids.production_id
```

File: addons/mrp_subcontracting/models/stock_picking.py [hunk 13]
```diff
@@ -18,96 +19,107 @@ def _compute_display_action_record_components(self):
...
-        action = self.env.ref('stock.stock_move_line_action').read()[0]
-        action['context'] = {}
-        action['domain'] = [('id', 'in', subcontracted_move_lines.ids)]
-        return action
+                # For concistency, set the date on production move before the date
+                # on picking. (Tracability report + Product Moves menu item)
+                minimum_date = min(picking.move_line_ids.mapped('date'))
+                production_moves = subcontracted_production.move_raw_ids | subcontracted_production.move_finished_ids
+                production_moves.write({'date': minimum_date - timedelta(seconds=1)})
+                production_moves.move_line_ids.write({'date': minimum_date - timedelta(seconds=1)})
```
Comment (sle-odoo) [inline addons/mrp_subcontracting/models/stock_picking.py]: noice

File: addons/mrp_subcontracting/models/stock_picking.py [hunk 14]
```diff
@@ -18,96 +19,107 @@ def _compute_display_action_record_components(self):
...
-            action = self.env.ref('mrp.act_mrp_product_produce').read()[0]
-            action['context'] = dict(self.env.context, active_id=mo.id)
-            return action
+        for move in self.move_lines:
+            if not move._has_tracked_subcontract_components():
+                continue
+            production = move.move_orig_ids.production_id
+            if not production or production.state in ('done', 'to_close'):
+                continue
+            return move._action_record_components()
```
Comment (sle-odoo) [inline addons/mrp_subcontracting/models/stock_picking.py]: allezz

File: addons/mrp_subcontracting/wizard/mrp_product_produce.py [hunk 15]
```diff
@@ -0,0 +1,85 @@
+# -*- coding: utf-8 -*-
+# Part of Odoo. See LICENSE file for full copyright and licensing details.
+
+from odoo import fields, models
+from odoo.tools.float_utils import float_compare, float_is_zero
+
+class MrpProductProduce(models.TransientModel):
+    _inherit = 'mrp.product.produce'
+
+    subcontract_move_id = fields.Many2one('stock.move')
```
Comment (sle-odoo) [inline addons/mrp_subcontracting/wizard/mrp_product_produce.py]: # stock move from the subcontract picking

File: addons/mrp_subcontracting/wizard/mrp_product_produce.py [hunk 16]
```diff
@@ -0,0 +1,85 @@
...
+
+    def continue_production(self):
+        action = super(MrpProductProduce, self).continue_production()
+        action['context'] = dict(action['context'], default_subcontract_move_id=self.subcontract_move_id.id)
+        return action
+
+    def _generate_produce_lines(self):
```
Comment (sle-odoo) [inline addons/mrp_subcontracting/wizard/mrp_product_produce.py]: add a comment
When the user did not have to use the produce wizard, this method is used to appropriately set the lot_produced_id and appropriately create a raw stock move line.

File: addons/mrp_subcontracting/wizard/mrp_product_produce.py [hunk 17]
```diff
@@ -0,0 +1,85 @@
...
+        self.ensure_one()
+        moves = (self.move_raw_ids | self.move_finished_ids).filtered(
+            lambda move: move.state not in ('done', 'cancel')
+        )
+        for move in moves:
+            qty_to_consume = self._prepare_component_quantity(move, self.qty_producing)
+            line_values = self._generate_lines_values(move, qty_to_consume)
+            self.env['mrp.product.produce.line'].create(line_values)
+
+    def _update_finished_move(self):
```
Comment (sle-odoo) [inline addons/mrp_subcontracting/wizard/mrp_product_produce.py]: after producing, set the move line on the subcontract picking

File: addons/mrp_subcontracting/wizard/mrp_product_produce.py [hunk 18]
```diff
@@ -0,0 +1,85 @@
...
+            self.env['stock.move.line'].create({
+                'move_id': self.subcontract_move_id.id,
+                'picking_id': self.subcontract_move_id.picking_id.id,
+                'product_id': self.product_id.id,
+                'location_id': self.subcontract_move_id.location_id.id,
+                'location_dest_id': self.subcontract_move_id.location_dest_id.id,
+                'product_uom_qty': 0,
+                'product_uom_id': self.product_uom_id.id,
+                'qty_done': quantity,
+                'lot_id': self.finished_lot_id and self.finished_lot_id.id,
```
Comment (sle-odoo) [inline addons/mrp_subcontracting/wizard/mrp_product_produce.py]: maybe just unlink all the reserved move lines instead of trying to synchronize it?

**Comment by sle-odoo** [comment]
robodoo r+ rebase-ff

**Comment by sle-odoo** [comment]
robodoo rebase-merge

**Pull request closed without merging**